### PR TITLE
Refresh quickly when no nodes in the pool

### PR DESCRIPTION
**Pull request by mdaudali**
## General
**Before this PR**:
If you ever end up in a situation where you have no clients in the pool, then you'll have at least 2 minutes of outage (or whatever your configured refresh time is) until you next refresh and reload the server list.

**After this PR**:

==COMMIT_MSG==
It's unnecessary to wait this long - when you have no clients, you have an outage for any endpoint that uses Atlas.
Instead, we implement a eager refresh when the current client pool is empty after the last refresh. We use [full jitter](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/) with exponential backoff to determine the next refresh, capping at 64 seconds (2^7)
==COMMIT_MSG==

**Priority**: P1

**Concerns / possible downsides (what feedback would you like?)**:
Is this too fast? If C* is down we'll immediately load the config, which _should_ allow us to make progress practically immediately, in the "normal" case.

Should we allow fast refresh to be configurable in whether we enable this or not?
**Is documentation needed?**:
No
## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:
No
**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:
No
**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:
Yes
**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:
No
**Does this PR need a schema migration?**
No
## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:
N/A
**What was existing testing like? What have you done to improve it?**:
More tests
**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:
N/A
**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:
N/A
## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:
We recover from 0 clients in pool quickly
**Has the safety of all log arguments been decided correctly?**:
Yes?
**Will this change significantly affect our spending on metrics or logs?**:
No
**How would I tell that this PR does not work in production? (monitors, etc.)**:
We never recover, or we kill Cassandra
**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:
Rollback
**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:
@jeremyk-91 
## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:
No
**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:
No
**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:
No, I don't think so
## Development Process
**Where should we start reviewing?**:
CCPI
**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:
N/A
**Please tag any other people who should be aware of this PR**:
@jeremyk-91 
@zpear 
@sverma30

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 1]
```diff
@@ -245,29 +250,38 @@ private CassandraClientPoolImpl(
...
+        }
+
+        if (getCurrentPools().isEmpty()) {
+            log.error(
+                    "There are no pools remaining after refreshing and validating pools. Scheduling the next refresh"
+                            + " very soon to avoid an extended downtime.",
+                    SafeArg.of("attempt", attempt));
+            int maxShift = Math.min(MAX_ATTEMPTS_BEFORE_CAPPING_BACKOFF, attempt);
+            refreshPoolFuture = refreshDaemon.schedule(
+                    () -> runAndScheduleNextRefresh(attempt + 1), RANDOM.nextInt(1 << maxShift), TimeUnit.SECONDS);
```
Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java]:
```suggestion
                    () -> runAndScheduleNextRefresh(attempt + 1), ThreadLocalRandom.current().nextInt(1 << maxShift), TimeUnit.SECONDS);
```

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 2]
```diff
@@ -245,29 +250,38 @@ private CassandraClientPoolImpl(
...
+                    t);
+        }
+
+        if (getCurrentPools().isEmpty()) {
+            log.error(
+                    "There are no pools remaining after refreshing and validating pools. Scheduling the next refresh"
+                            + " very soon to avoid an extended downtime.",
+                    SafeArg.of("attempt", attempt));
+            int maxShift = Math.min(MAX_ATTEMPTS_BEFORE_CAPPING_BACKOFF, attempt);
+            refreshPoolFuture = refreshDaemon.schedule(
```
Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java]: does `refreshPoolFuture` need to be `volatile` for safe publication and accesses?

Comment (mdaudali) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java]: Yeah, have that and a few other changes to push up still :) Thanks for a look over, though!

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 3]
```diff
@@ -245,29 +248,49 @@ private CassandraClientPoolImpl(
...
+        scheduleNextRefresh(attempt);
+    }
+
+    // TODO(mdaudali): We can inline this into runAndScheduleNextRefresh once #6753 is merged
+    private void scheduleNextRefresh(int attempt) {
+        if (getCurrentPools().isEmpty()) {
+            int maxShift = Math.min(MAX_ATTEMPTS_BEFORE_CAPPING_BACKOFF, attempt);
+
+            // Caps out at 2^7 * 1000 = 64000
+            int millisTillNextRefresh = Math.max(1, ThreadLocalRandom.current().nextInt((1 << maxShift) * 1000));
```
Comment (mdaudali) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java]: min clamp at 1 is mostly to not require rewriting a bunch of tests that verify a single refresh, which gets violated if you generate 0 (this happens on the first load, + random luck which happened more often when the state space was smaller (seconds, not millis)

Comment (jeremyk-91) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java]: yeah normally I'd use `Duration.SECONDS.toMillis()` or something like that? But given time limit, but ok

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 4]
```diff
@@ -288,7 +311,9 @@ public void shutdown() {
     }
 
     private void cancelRefreshPoolTaskIfRunning() {
-        if (refreshPoolFuture != null) {
+        // It's possible for the cancelled refreshPoolFuture to schedule another task before terminating. If it does
+        // schedule another task, we can just cancel that one too.
+        while (refreshPoolFuture != null && !refreshPoolFuture.isCancelled()) {
```
Comment (jeremyk-91) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:319]: nice

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 5]
```diff
@@ -245,29 +248,49 @@ private CassandraClientPoolImpl(
...
+                    "Failed to refresh Cassandra KVS pool."
+                            + " Extended periods of being unable to refresh will cause perf degradation.",
+                    t);
+        }
+
+        scheduleNextRefresh(attempt);
+    }
+
+    // TODO(mdaudali): We can inline this into runAndScheduleNextRefresh once #6753 is merged
+    private void scheduleNextRefresh(int attempt) {
```
Comment (jeremyk-91) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java]: consecutivelyFailedAttempts?

**Review by jeremyk-91 — APPROVED** [review]
Approach is good. Some minorish quality bits but given priority not blocking.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java [hunk 6]
```diff
@@ -288,7 +311,9 @@ public void shutdown() {
     }
 
     private void cancelRefreshPoolTaskIfRunning() {
-        if (refreshPoolFuture != null) {
+        // It's possible for the cancelled refreshPoolFuture to schedule another task before terminating. If it does
+        // schedule another task, we can just cancel that one too.
+        while (refreshPoolFuture != null && !refreshPoolFuture.isCancelled()) {
             refreshPoolFuture.cancel(true);
```
Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:320]: Do we need to null out `refreshPoolFuture`?

**Pull request merged**